File: KVODelegateObjCTests/PersonB.m
//
//  Person.m
//  KVODelegate
//
//  Created by Ian Gregory on 16-03-2017.
//  Copyright © 2017 ThatsJustCheesy. All rights reserved.
//

#import "PersonB.h"

@implementation PersonB

- (NSString *)fullName {
    return [NSString stringWithFormat:@"%@ %@", self.firstName, self.lastName];
}

- (void)setFullName:(NSString *)fullName {
    NSUInteger spaceLoc = [fullName rangeOfString:@" "].location;
    _firstName = [fullName substringToIndex:spaceLoc];
    _lastName = [fullName substringFromIndex:spaceLoc + 1];
}

+ (void)configKVONotificationDelegate:(KVONotificationDelegate *)delegate {
    [delegate key:@"fullName" dependsUponKeyPaths:@[@"firstName", @"lastName"]];
    [delegate key:@"firstName" dependsUponKeyPath:@"fullName"];
    [delegate key:@"lastName" dependsUponKeyPath:@"fullName"];
}
KVO_DELEGATE_HANDLES_keyPathsForValuesAffectingValueForKey;

@end

File: KVODelegateSwiftTests/PersonB.swift
//
//  PersonB.swift
//  KVODelegate
//
//  Created by Ian Gregory on 17-03-2017.
//  Copyright © 2017 ThatsJustCheesy. All rights reserved.
//

import Foundation
import KVODelegate

class PersonB: NSObject, KVONotificationDelegator {
    
    @objc dynamic var firstName = "", lastName = "", address = "", postalCode = ""
    
    @objc dynamic var fullName: String {
        get {
            return firstName + " " + lastName
        }
        set {
            let spaceLoc = newValue.range(of: " ")!
            firstName = String(newValue[..<spaceLoc.lowerBound])
            lastName = String(newValue[newValue.index(after: spaceLoc.lowerBound)...])
        }
    }
    
    @objc dynamic var creditCardNumbers: [Int] = []
    
    class func configKVONotificationDelegate(_ delegate: KVONotificationDelegate) {
        delegate.key("fullName", dependsUponKeyPaths: ["firstName", "lastName"])
        delegate.key("firstName", dependsUponKeyPath: "fullName")
        delegate.key("lastName", dependsUponKeyPath: "fullName")
    }
    
    // Forcing the user to manually override is kinda clumsy, but it's the only way in Swift
    override class func keyPathsForValuesAffectingValue(forKey key: String) -> Set<String> {
        return KVONotificationDelegate(forClass: self).keyPathsForValuesAffectingValue(forKey: key)
    }
    
}
